buildSrc/subprojects/dependency-modules/src/main/kotlin/gradlebuild/dependency-modules.gradle.kts

```diff
@@ -1,5 +1,278 @@
+        withLibraryDependencies<DependencyRemovalByGroupRule>(libs.maven3PluginApi, setOf("org.sonatype.sisu"))
+        withLibraryDependencies<DependencyRemovalByGroupRule>(libs.maven3SettingsBuilder, setOf("org.sonatype.sisu"))
+
+        // We don't need the extra annotations provided by j2objc
+        withLibraryDependencies<DependencyRemovalByNameRule>(libs.googleHttpClient, setOf("j2objc-annotations"))
+
+        // Read capabilities declared in capabilities.json
+        readCapabilitiesFromJson()
+
+        withModule<ReplaceCglibNodepWithCglibRule>("org.spockframework:spock-core")
+        // Prevent Spock from pulling in Groovy and third-party dependencies - see https://github.com/spockframework/spock/issues/899
+        withLibraryDependencies<DependencyRemovalByNameRule>("org.spockframework:spock-core",
+            setOf("groovy-groovysh", "groovy-json", "groovy-macro", "groovy-nio", "groovy-sql", "groovy-templates", "groovy-test", "groovy-xml"))
+        withLibraryDependencies<DependencyRemovalByNameRule>("cglib:cglib", setOf("ant"))
+
+        // asciidoctorj depends on a lot of stuff, which causes `Can't create process, argument list too long` on Windows
+        withLibraryDependencies<DependencyRemovalByNameRule>("org.gradle:sample-discovery", setOf("asciidoctorj", "asciidoctorj-api"))
+
+        withModule<DowngradeXmlApisRule>("jaxen:jaxen")
+        withModule<DowngradeXmlApisRule>("jdom:jdom")
+        withModule<DowngradeXmlApisRule>("xalan:xalan")
+        withModule<DowngradeXmlApisRule>("jaxen:jaxen")
+
+        // We only need "failureaccess" of Guava's dependencies
+        withLibraryDependencies<KeepDependenciesByNameRule>("com.google.guava:guava", setOf("failureaccess"))
+
+        // Test dependencies - minify: remove unused transitive dependencies
+        withLibraryDependencies<DependencyRemovalByNameRule>("org.gradle.org.littleshoot:littleproxy",
+            setOf("barchart-udt-bundle", "guava", "commons-cli"))
+
+        // TODO: Gradle profiler should use the bundled tooling API.
+        //   This should actually be handled by conflict resolution, though it doesn't seem to work.
+        //   See https://github.com/gradle/gradle/issues/12002.
+        withLibraryDependencies<DependencyRemovalByNameRule>("org.gradle.profiler:gradle-profiler",
+            setOf("gradle-tooling-api"))
+    }
+}
+
+
+fun applyAutomaticUpgradeOfCapabilities() {
+    configurations.all {
+        resolutionStrategy.capabilitiesResolution.all {
+            selectHighestVersion()
+        }
+    }
+}
+
+fun readCapabilitiesFromJson() {
+    val extra = gradle.rootProject.extra
+    val capabilities: List<CapabilitySpec>
+    if (extra.has("capabilities")) {
+        @Suppress("unchecked_cast")
+        capabilities = extra["capabilities"] as List<CapabilitySpec>
+    } else {
+        val capabilitiesFile = gradle.rootProject.layout.projectDirectory.file("gradle/dependency-management/capabilities.json").asFile
```

Comment:
Looks like a good candidate for `providers.fileContents`

Reply:
A more severe issue is that this code here is using `ext` as a cache. This should be replaced by a build service - if we want to keep the "capabilities.json" around. I have mixed feelings about that one as it is a very specific solution. I would prefer is we would use more "out of the box" APIs here. (Then we won't need to read a file at all.)

Reply:
The idea of this file is to prototype an external form for those rules. I think users would appreciate that. Code isn't nice and we already have a number of comments from people saying "no way I'm going to write those component metadata rules".

Reply:
:+1:

I'll create a follow up issue to change this code here to use a build service then.

Reply:
https://github.com/gradle/gradle-private/issues/3141